src/data_source/sql.rs

```diff
@@ -608,134 +649,408 @@ impl VersionedDataSource for SqlDataSource {
 
-    type LeafRange<'a, R> = BoxStream<'a,  QueryResult<LeafQueryData<Types, I>>>
+    type LeafRange<'a, R> = BoxStream<'static, QueryResult<LeafQueryData<Types, I>>>
     where
         Self: 'a,
         R: RangeBounds<usize> + Send;
-    type BlockRange<'a, R>= BoxStream<'a,  QueryResult<BlockQueryData<Types>>>
+    type BlockRange<'a, R>= BoxStream<'static, QueryResult<BlockQueryData<Types>>>
     where
         Self: 'a,
         R: RangeBounds<usize> + Send;
 
-    async fn get_leaf<ID>(&self, _id: ID) -> QueryResult<LeafQueryData<Types, I>>
+    async fn get_leaf<ID>(&self, id: ID) -> QueryResult<LeafQueryData<Types, I>>
     where
         ID: Into<LeafId<Types, I>> + Send + Sync,
     {
-        todo!()
+        let (where_clause, param): (&str, Box<dyn ToSql + Send + Sync>) = match id.into() {
+            ResourceId::Number(n) => ("height = $1", Box::new(n as i64)),
+            ResourceId::Hash(h) => ("hash = $1", Box::new(h.to_string())),
+        };
+        let query = format!("SELECT leaf, qc FROM leaf WHERE {where_clause}");
+        let row = self
+            .client
+            .query_opt(&query, &[&*param])
+            .await
+            .map_err(|err| QueryError::Error {
+                message: err.to_string(),
+            })?
+            .context(NotFoundSnafu)?;
+        parse_leaf(row)
     }
-    async fn get_block<ID>(&self, _id: ID) -> QueryResult<BlockQueryData<Types>>
+
+    async fn get_block<ID>(&self, id: ID) -> QueryResult<BlockQueryData<Types>>
     where
         ID: Into<BlockId<Types>> + Send + Sync,
     {
-        todo!()
+        let (where_clause, param): (&str, Box<dyn ToSql + Send + Sync>) = match id.into() {
+            ResourceId::Number(n) => ("h.height = $1", Box::new(n as i64)),
+            ResourceId::Hash(h) => ("h.hash = $1", Box::new(h.to_string())),
+        };
+        // ORDER BY h.height ASC ensures that if there are duplicate blocks, we return the first one.
+        let query = format!(
+            "SELECT {BLOCK_COLUMNS}
+              FROM header AS h
+              JOIN payload AS p ON h.height = p.height
+              WHERE {where_clause}
+              ORDER BY h.height ASC
+              LIMIT 1"
+        );
+        let row = self
+            .client
+            .query_opt(&query, &[&*param])
+            .await
+            .map_err(|err| QueryError::Error {
+                message: err.to_string(),
+            })?
+            .context(NotFoundSnafu)?;
+        parse_block(row)
     }
 
-    async fn get_leaf_range<R>(&self, _range: R) -> QueryResult<Self::LeafRange<'_, R>>
+    async fn get_leaf_range<R>(&self, range: R) -> QueryResult<Self::LeafRange<'_, R>>
     where
         R: RangeBounds<usize> + Send,
     {
-        todo!()
+        let (where_clause, params) = bounds_to_where_clause(range, "height");
+        let query = format!("SELECT leaf, qc FROM leaf{where_clause} ORDER BY height ASC");
```

Comment:
Nit: looks a bit weird without the space between `leaf{where_clause}`.

Reply:
Fixed in [996c49b](https://github.com/EspressoSystems/hotshot-query-service/pull/249/commits/996c49bb55f47ff7f530bc3813239f1ee41fe11f)